File: swift/Sources/DSA/Algorithms/MergeSort.swift
/// Merge sort algorithm.
/// Time: O(n log n) all cases
/// Space: O(n)
/// Stable: Yes
public enum MergeSort {
    /// Sort an array in-place.
    public static func sort<T: Comparable>(_ array: inout [T], by comparator: ((T, T) -> Bool)? = nil) {
        array = mergeSort(array, by: comparator)
    }

    /// Return a sorted copy of the array.
    public static func sorted<T: Comparable>(_ array: [T], by comparator: ((T, T) -> Bool)? = nil) -> [T] {
        mergeSort(array, by: comparator)
    }

    private static func mergeSort<T: Comparable>(_ array: [T], by comparator: ((T, T) -> Bool)?) -> [T] {
        guard array.count > 1 else { return array }

        let compare = comparator ?? { $0 < $1 }
        let mid = array.count / 2
        let left = mergeSort(Array(array[..<mid]), by: comparator)
        let right = mergeSort(Array(array[mid...]), by: comparator)

        return merge(left, right, by: compare)
    }

    private static func merge<T>(_ left: [T], _ right: [T], by compare: (T, T) -> Bool) -> [T] {
        var result: [T] = []
        result.reserveCapacity(left.count + right.count)

        var i = 0
        var j = 0

        while i < left.count && j < right.count {
            if compare(left[i], right[j]) || (!compare(right[j], left[i]) && i < j) {
                result.append(left[i])
                i += 1
            } else {
                result.append(right[j])
                j += 1
            }
        }

        result.append(contentsOf: left[i...])
        result.append(contentsOf: right[j...])

        return result
    }
}

File: kotlin/src/main/kotlin/algorithms/MergeSort.kt
package algorithms

/**
 * Merge sort algorithm.
 * Time: O(n log n) all cases
 * Space: O(n)
 * Stable: Yes
 */
object MergeSort {
    /**
     * Sort a mutable list in-place.
     */
    fun <T : Comparable<T>> sort(list: MutableList<T>, comparator: Comparator<T>? = null) {
        val sorted = mergeSort(list, comparator)
        for (i in sorted.indices) {
            list[i] = sorted[i]
        }
    }

    /**
     * Return a sorted copy of the list.
     */
    fun <T : Comparable<T>> sorted(list: List<T>, comparator: Comparator<T>? = null): List<T> =
        mergeSort(list, comparator)

    private fun <T : Comparable<T>> mergeSort(list: List<T>, comparator: Comparator<T>?): List<T> {
        if (list.size <= 1) return list

        val mid = list.size / 2
        val left = mergeSort(list.subList(0, mid), comparator)
        val right = mergeSort(list.subList(mid, list.size), comparator)

        return merge(left, right, comparator)
    }

    private fun <T : Comparable<T>> merge(left: List<T>, right: List<T>, comparator: Comparator<T>?): List<T> {
        val result = mutableListOf<T>()
        var i = 0
        var j = 0

        while (i < left.size && j < right.size) {
            if (compare(left[i], right[j], comparator) <= 0) {
                result.add(left[i++])
            } else {
                result.add(right[j++])
            }
        }

        while (i < left.size) result.add(left[i++])
        while (j < right.size) result.add(right[j++])

        return result
    }

    private fun <T : Comparable<T>> compare(a: T, b: T, comparator: Comparator<T>?): Int =
        comparator?.compare(a, b) ?: a.compareTo(b)
}
